Introduce static elementary types to use

lastmjs opened the pull request:
### Checklist
- [x] Code compiles correctly
- [ ] All tests are passing (except linter, still need help on that)
- [x] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary (I don't believe there is anything to change here)
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### Description

Completing https://github.com/ethereum/solidity/issues/4502

lastmjs commented:
I'm kind of confused by the code coverage failing here, can someone point me in the right direction?

chriseth commented:
Thanks for your contribution! While reviewing, @ekpyron noted that the static objects have to be references to shared pointers, because otherwise `shared_from_this` would not work. In general, we should strive to disallow bare objects of types at all, which is of course out of scope for this PR.

One way to both correct the problem that these objects here have to be shared pointers and go into the direction of disallowing bare objects would be the following:

Instead of a struct containing these static members, add public static functions to the types which have a static local variable of shared pointer type that return a reference to the object.

So it would e.g. be:

```
class IntegerType
{
...
static IntegerType& address() { static shared_ptr<IntegerType> addr(new IntegerType(160, Modifier::Address)); return *addr; }
}
```

Using functions also has the advantage that we get around the static initialization order problem.

lastmjs commented:
I have done your bidding. Let me know what you think.

ekpyron commented:
This looks good, but unfortunately, while you were working on this PR, #4887 was merged, which is now in conflict with this PR - do you think you could rebase this PR and adjust it to these latest changes in ``develop``? Instead of ``IntegerType::address()`` it will now have to be ``AddressType::address()``.

ekpyron commented:
Rebasing will probably also fix the external test runs, which break due to breaking changes in separately merged PRs.

lastmjs commented:
I'm on it

lastmjs commented:
Ready for final review

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -319,6 +319,8 @@ class Type: private boost::noncopyable, public std::enable_shared_from_this<Type
 class AddressType: public Type
 {
 public:
+    static AddressType& address() { static std::shared_ptr<AddressType> addr(new AddressType()); return *addr; }
```
Uses spaces for indentation.

chriseth replied:
Also please don't use `new`. Use `make_shared` instead.

lastmjs replied:
Am I not using spaces for indentation? My editor is set to use 4 spaces for tabs, I'm trying to change that line but I think it's already spaces

chriseth replied:
I'm sorry, this was a little unclear: You should use tabs everywhere, but these are spaces.

lastmjs replied:
Got it

chriseth reviewed (changes requested):
Please do the small changes and then squash everything into a single commit. The first commit still contains merge conflict markers.

ekpyron commented:
Unfortunately, this will need to be rebased again, if #4911 is merged before it.

lastmjs committed:
fixing rebase conflicts

add ElementaryTypes::BytesMemory and ElementaryTypes::StringMemory

fix rebase conflicts

fixing rebase conflicts

use make_shared instead of new

fix tabs and StateMutability parameter

create address and addressType functions, fix spaces in comment

fix typo and switch nonpayable to payable

fix spaces

fix comment again

lastmjs commented:
It looks like the `AddressType` constructor has a new `StateMutability` parameter. What should I pass in here when creating the shared pointer?

```c++
static AddressType& address() { static std::shared_ptr<AddressType> addr(std::make_shared<AddressType>()); return *addr; }
```

It seems like the `address` function should take the `StateMutability` as a parameter. In that case, what do I pass in when calling `address`?

lastmjs commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2131,7 +2131,7 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 				(memberName == "value" ? " - did you forget the \"payable\" modifier?" : ".");
 		if (exprType->category() == Type::Category::Contract)
-			for (auto const& addressMember: AddressType(StateMutability::Payable).nativeMembers(nullptr))
+			for (auto const& addressMember: AddressType::address(StateMutability::Payable).nativeMembers(nullptr))
```
I have no idea what the `StateMutability` parameter should be here

lastmjs commented on libsolidity/ast/Types.h:
```diff
@@ -1,18 +1,18 @@
 /*
-    This file is part of solidity.
+	This file is part of solidity.
```
I just blanket-converted all spaces to tabs in this file, did we also want the comment to use tabs?

chriseth replied:
This is a very old file, please just keep existing lines in comments.

chriseth commented:
Please create two functions: One just called `address` with state mutability "nonpayable" and one called `addressPayable` with state mutability 'payable'.

lastmjs commented:
When should I use each? There are a couple of instances in this pull request where I need to decide

chriseth commented:
The current code should already have them.

chriseth commented:
Can you please squash the commits into one?

lastmjs commented:
I'm not sure what the test failures are, tests look fine on my machine...oh wait, looks like some package.json versioning issue
